Rust/src/lib.rs

```diff
@@ -51,61 +56,67 @@ mod tests {
-    }
+        let tests = vec![
+            Test {
+                input: AfricanParrot { parrot: ParrotBase{ number_of_coconuts: 1, voltage: 0.0, nailed: false}},
+                output: 3.0
+            },
+            Test {
+                input: AfricanParrot { parrot: ParrotBase{ number_of_coconuts: 2, voltage: 0.0, nailed: false }},
+                output: 0.0
+            },
+            Test {
+                input: AfricanParrot { parrot: ParrotBase{ number_of_coconuts: 0, voltage: 0.0, nailed: false}},
+                output: 12.0
+            },
+        ];
 
-    #[test]
-    fn african_parrot_speed_with_no_coconut() {
-        let parrot = Parrot { parrot_type: "african_parrot",
-                              number_of_coconuts: 0,
-                              voltage: 0.0,
-                              nailed: false };
-        assert_eq!(parrot.speed().unwrap(), 12.0);
-    }
-    #[test]
-    fn nailed_norwegian_blue_parrot() {
-        let parrot = Parrot { parrot_type: "norwegian_blue_parrot",
-                              number_of_coconuts: 0,
-                              voltage: 1.5,
-                              nailed: true };
-        assert_eq!(parrot.speed().unwrap(), 0.0);
-    }
-    #[test]
-    fn not_nailed_norwegian_blue_parrot() {
-        let parrot = Parrot { parrot_type: "norwegian_blue_parrot",
-                              number_of_coconuts: 0,
-                              voltage: 1.5,
-                              nailed: false };
-        assert_eq!(parrot.speed().unwrap(), 18.0);
+        for test in tests {
+             let result = test.input;
+             assert_eq!(result.speed(), test.output);
+        }
+    
     }
+
     #[test]
-    fn not_nailed_norwegian_blue_parrot_with_high_voltage() {
-        let parrot = Parrot { parrot_type: "norwegian_blue_parrot",
-                              number_of_coconuts: 0,
-                              voltage: 4.0,
-                              nailed: false };
-        assert_eq!(parrot.speed().unwrap(), 24.0);
+    fn norwegian_blue_parrot_speed(){
+        struct Test {
+            input: NorwegianBlueParrot,
+            output: f32,
+        }
+
+        let tests = vec![
+            Test {
+                input:  NorwegianBlueParrot { parrot: ParrotBase{ number_of_coconuts: 2, voltage: 0.0, nailed: true }},
+                output: 0.0
+            },
+            Test {
+                input: NorwegianBlueParrot { parrot: ParrotBase{ number_of_coconuts: 0, voltage: 1.5, nailed: false }},
+                output: 18.0
+            },
+            Test {
+                input: NorwegianBlueParrot { parrot: ParrotBase{ number_of_coconuts: 0, voltage: 4.0, nailed: false }},
+                output: 24.0
+            },
+        ];
+
+        for test in tests {
+             let result = test.input;
+             assert_eq!(result.speed(), test.output);
+        }
+    
     }
```

Comment:
I guess the only downside of this as opposed to using traits is that you can't dry up the test runner/iterator. Although I think you'd have to use a macro as otherwise the compiler wouldn't know what kind of parrot it was dealing with.

In these simple cases it's arguable whether it's worth the hassle though, I decided to do neither and now regret it after seeing your's and Darren's solutions :see_no_evil: